consensus/src/block_storage/block_store.rs

```diff
@@ -274,6 +282,9 @@ impl BlockStore {
             return Ok(existing_block);
         }
         let executed_block = self.execute_block(block)?;
+        // ensure local time past the block time
+        let block_time = Duration::from_micros(executed_block.timestamp_usecs());
+        self.time_service.wait_until(block_time);
```

Comment:
We know that the wait_time is bounded by the round duration, however it is not obvious since that check occurs outside of this function.  A naive caller could wait indefinitely.  

It would be helpful to add this assumption to the function comment that the block must have a valid timestamp that is within the current round duration.  If someone ever removes that check, it could open up a validator for infinite waiting from a rogue proposal.

Reply:
It might also be worth logging if the wait is a lot longer than expected.  For instance if the wait is > 1 minute, something is probably really wrong either with this node's clock or the quorum's clock.